src/Processors/Formats/Impl/AvroRowInputFormat.cpp

```diff
@@ -438,6 +434,80 @@ AvroDeserializer::DeserializeFn AvroDeserializer::createDeserializeFn(const avro
                         target_type->getName());
                 }
             }
+
+            if (target.isVariant())
+            {
+                const auto & variant_type = assert_cast<const DataTypeVariant &>(*target_type);
+                const auto & nested_types = variant_type.getVariants();
+
+                using AvroUnionIndex = size_t;
+                std::map<AvroUnionIndex, ColumnVariant::Discriminator> union_index_to_global_discriminator;
+                std::vector<DeserializeFn> nested_deserializers;
+                nested_deserializers.reserve(root_node->leaves());
+
+                bool union_has_null = false;
+                for (size_t i = 0; i != root_node->leaves(); ++i)
+                {
+                    const auto & avro_node = root_node->leafAt(static_cast<int>(i));
+                    if (avro_node->type() == avro::AVRO_NULL)
+                    {
+                        union_has_null = true;
+                        nested_deserializers.emplace_back();
+                        union_index_to_global_discriminator.insert_or_assign(i, ColumnVariant::NULL_DISCRIMINATOR);
+                        continue;
+                    }
+                    const auto variant = AvroSchemaReader::avroNodeToDataType(avro_node);
+                    nested_deserializers.emplace_back(createDeserializeFn(avro_node, variant));
+                    union_index_to_global_discriminator.insert_or_assign(i, variant_type.tryGetVariantDiscriminator(variant->getName()).value());
```

Comment:
It's not necessary that we will have this variant. It will never happen when schema inference is used, but the user can specify custom `Variant` type. Let's throw an exception that Variant and union are not compatible in this case